sympy/matrices/matrices.py

```diff
@@ -1581,6 +1583,11 @@ def _matrix_pow_by_recursion(self, num):
     def __pow__(self, num):
         if not self.is_square:
             raise NonSquareMatrixError()
+        num = sympify(num)
+        if not num.is_integer and (num >= 0) == True and not num % 1 == 0 and self.det() == 0:
```

Comment:
The determinant is computed in connection with Jordan blocks in `_matrix_pow_by_jordan_blocks`. Computing it twice is not an improvement.